modules/stats/metrics.go

```diff
@@ -0,0 +1,115 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package stats
+
+import (
+	"go.uber.org/fx/service"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	//TagModule is module tag for metrics
+	TagModule = "module"
+	// TagType is either request or response
+	TagType = "type"
+	// TagStatus is status of the request
+	TagStatus = "status"
+)
+
+var httpMethods = []string{"GET", "HEAD", "PUT", "POST", "DELETE", "CONNECT", "OPTIONS", "TRACE", "PATCH"}
+
+// HTTPTags creates metrics scope with defined tags
+var (
+	HTTPTags = map[string]string{
+		TagModule: "http",
+		TagType:   "request",
+	}
+
+	// RPCTags creates metrics scope with defined tags
+	RPCTags = map[string]string{
+		TagModule: "rpc",
+		TagType:   "request",
+	}
+
+	// TaskTags creates metrics scope with defined tags
+	TaskTags = map[string]string{
+		TagModule: "task",
+		TagType:   "request",
+	}
+)
+
+var (
+	// HTTPPanicCounter counts panics occurred in http
+	HTTPPanicCounter tally.Counter
+	// HTTPAuthFailCounter counts auth failures
+	HTTPAuthFailCounter tally.Counter
+	// HTTPMethodTimer is a turnaround time for http methods
+	HTTPMethodTimer map[string]tally.Timer
+	// HTTPStatusCountScope is a scope for http status
+	HTTPStatusCountScope tally.Scope
+
+	// RPCAuthFailCounter counts auth failures
+	RPCAuthFailCounter tally.Counter
+	// RPCHandleTimer is a turnaround time for rpc handler
+	RPCHandleTimer tally.Scope
+
+	// TaskExecutionCount counts number of executions
+	TaskExecutionCount tally.Counter
+	// TaskPublishCount counts number of tasks enqueued
+	TaskPublishCount tally.Counter
+	// TaskExecuteFail counts number of tasks failed to execute
+	TaskExecuteFail tally.Counter
+	// TaskPublishFail counts number of tasks failed to enqueue
+	TaskPublishFail tally.Counter
+	// TaskExecutionTime is a turnaround time for execution
+	TaskExecutionTime tally.Timer
+	// TaskPublishTime is a publish time for tasks
+	TaskPublishTime tally.Timer
+)
+
+// SetupHTTPMetrics allocates counters for necessary setup
+func SetupHTTPMetrics(host service.Host) {
+	HTTPPanicCounter = host.Metrics().Tagged(HTTPTags).Counter("panic")
+	HTTPAuthFailCounter = host.Metrics().Tagged(HTTPTags).Counter("http.auth.fail")
+	HTTPMethodTimer = make(map[string]tally.Timer)
+	for _, method := range httpMethods {
+		HTTPMethodTimer[method] = host.Metrics().Tagged(HTTPTags).Timer("http." + method + ".time")
+	}
+	HTTPStatusCountScope = host.Metrics().Tagged(HTTPTags)
+}
+
+// SetupRPCMetrics allocates counters for necessary setup
+func SetupRPCMetrics(host service.Host) {
+	RPCAuthFailCounter = host.Metrics().Tagged(RPCTags).Counter("rpc.auth.fail")
+	RPCHandleTimer = host.Metrics().Tagged(RPCTags)
+}
+
+// SetupTaskMetrics allocates counters for necessary setup
+func SetupTaskMetrics(host service.Host) {
```

Comment:
Why do we setup metrics for all modules in the metrics module? If someone wants do delete or update metrics they need to make change in their own local module, not in 2 places..